helix-core/src/main/java/org/apache/helix/cloud/event/CloudEventHandler.java

```diff
@@ -0,0 +1,92 @@
+package org.apache.helix.cloud.event;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CloudEventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(CloudEventHandler.class.getName());
+  @VisibleForTesting
+  protected List<CloudEventListener> _eventListeners = new ArrayList<>();
+  private PreEventHandlerCallback _preEventHandlerCallback = null;
+  private PostEventHandlerCallback _postEventHandlerCallback = null;
```

Comment:
Except us, nobody may hook this, right? It is to fine-grained to have explicit call pre/post.

We can make internal property of CallbackHanlder to differentiate them by a flag or enum type inside EventHanlderCallback class. It will reduce functions or declarations.

Reply:
Sorry for the confusion. This is for user to hook, they can implement them and register them to the event handler.
I think you are talking about the pre- and post- **callback** inside each listener. 
These two are different - conceptually they will be the first and the last **event listener** to be triggered in an event.
If it's not that clear maybe we can rename these two interfaces?

Reply:
Rename and simplify it with single class with different internal properties could be better. We dont have to get two separate class that invoke in the code.